rllib/connectors/action.py

```diff
@@ -0,0 +1,173 @@
+from typing import Any, List
+
+from ray.rllib.connectors.connector import (
+    Connector,
+    ConnectorContext,
+    ConnectorPipeline,
+    ActionConnector,
+    register_connector,
+    get_connector,
+)
+from ray.rllib.utils.annotations import DeveloperAPI
+from ray.rllib.utils.numpy import convert_to_numpy
+from ray.rllib.utils.spaces.space_utils import (
+    clip_action,
+    get_base_struct_from_space,
+    unsquash_action,
+    unbatch,
+)
+from ray.rllib.utils.typing import (
+    ActionConnectorDataType,
+    TrainerConfigDict,
+)
+
+
+@DeveloperAPI
+class ConvertToNumpyConnector(ActionConnector):
+    def __call__(self, ac_data: ActionConnectorDataType) -> ActionConnectorDataType:
+        assert isinstance(
+            ac_data.output, tuple
+        ), "Action connector requires PolicyOutputType data."
+
+        actions, states, fetches = ac_data.output
+        return ActionConnectorDataType(
+            ac_data.env_id,
+            ac_data.agent_id,
+            (convert_to_numpy(actions), convert_to_numpy(states), fetches),
```

Comment:
The elephant in the room is: many of these examples are simple, yet verbose wrappers around existing functionality. We already supply context for each of them. Sometimes they have additional state. The question is (assuming that we can move all state into context?) whether we're better off with a functional design like this:

```python
def connector(ctx: ConnectorContext, data ConnectorData): -> ConnectorData
    ...
```

that you can chain (e.g. as simple list of functions)

```
connectors = [convert_to_numpy, unbatch, my_custom_fct]
```

Reply:
kind of depends on how "hidden" connectors are. E.g. if I have a custom transform, I want to make it work quickly and not inherit a class, if possible. I guess it's also possible to have a `LambdaConnector` that takes in a user-defined function. Just thinking about how we can make this most attractive and flexible.

Reply:
Done adding LambdaAgentConnector and LambdaActionConnector. Simplified the implementation of a few existing connectors.